src/qibo/quantum_info/linalg_operations.py

```diff
@@ -0,0 +1,227 @@
+            + f"but it is {2**nqubits}.",
+        )
+
+    nqubits = int(nqubits)
+
+    statevector = bool(len(state.shape) == 1)
+
+    factor = 1 if statevector else 2
+    state = backend.np.reshape(state, factor * nqubits * (2,))
+
+    if statevector:
+        axes = 2 * [list(traced_qubits)]
+        rho = backend.np.tensordot(state, backend.np.conj(state), axes)
+        shape = 2 * (2 ** (nqubits - len(traced_qubits)),)
+
+        return backend.np.reshape(rho, shape)
+
+    order = tuple(sorted(traced_qubits))
+    order += tuple(i for i in range(nqubits) if i not in traced_qubits)
+    order += tuple(i + nqubits for i in order)
+    shape = 2 * (2 ** len(traced_qubits), 2 ** (nqubits - len(traced_qubits)))
+
+    state = backend.np.transpose(state, order)
+    state = backend.np.reshape(state, shape)
+
+    return backend.np.einsum("abac->bc", state)
+
+
+def matrix_exponentiation(
+    phase: Union[float, complex],
+    matrix,
+    eigenvectors=None,
+    eigenvalues=None,
+    backend=None,
+):
+    """Calculates the exponential of a matrix.
+
+    Given a ``matrix`` :math:`H` and a ``phase`` :math:`\\theta`,
+    it returns the exponential of the form
+
+    .. math::
+        \\exp\\left(-i \\, \\theta \\, H \\right) \\, .
+
+    If the ``eigenvectors`` and ``eigenvalues`` are given, the matrix diagonalization
+    is used for the exponentiation.
+
+    Args:
+        phase (float or complex): phase that multiplies the matrix.
+        matrix (ndarray): matrix to be exponentiated.
+        eigenvectors (ndarray, optional): _if not ``None``, eigenvectors are used
+            to calculate ``matrix`` exponentiation as part of diagonalization.
+            Must be used together with ``eigenvalues``. Defaults to ``None``.
+        eigenvalues (ndarray, optional): if not ``None``, eigenvalues are used
+            to calculate ``matrix`` exponentiation as part of diagonalization.
+            Must be used together with ``eigenvectors``. Defaults to ``None``.
+        backend (:class:`qibo.backends.abstract.Backend`, optional): backend
+            to be used in the execution. If ``None``, it uses
+            :class:`qibo.backends.GlobalBackend`. Defaults to ``None``.
+
+    Returns:
+        ndarray: matrix exponential of :math:`-i \\, \\theta \\, H`.
+    """
+    backend = _check_backend(backend)
+    backend_class_name = backend.__class__.__name__
+
+    is_sparse = backend.is_sparse(matrix)
+
+    if eigenvectors is None or is_sparse:
+        matrix *= -1j * phase
+
+        if backend_class_name == "TensorflowBackend":
+            return backend.tf.linalg.expm(matrix)
+
+        if backend_class_name == "PyTorchBackend":
+            return backend.np.linalg.matrix_exp(matrix)
+
+        if is_sparse:
+            expm = expm_sparse_scipy
+        elif backend_class_name in ["CupyBackend", "CuQuantumBackend"]:
+            from cupyx.scipy.linalg import expm as expm_cupy  # pylint: disable=E0401
```

Comment:
If a function explicitly references individual backends, then it should be part of the backend definition.
Backends may change, from content to even the name. And we should be free to do so without scanning the whole Qibo for explicit mentions.

If you want to expose a function in this module, to have the feature together with other similar ones, then keep the `calculate_matrix_exp` in the backends, and call that one inside here.